packages/@expo/cli/src/start/server/middleware/CorsMiddleware.ts

```diff
@@ -2,27 +2,49 @@ import type { ExpoConfig } from '@expo/config';
 
 import type { ServerRequest, ServerResponse } from './server.types';
 
-const DEFAULT_ALLOWED_CORS_HOSTNAMES = [
-  'localhost',
+const DEFAULT_ALLOWED_CORS_HOSTS = [
   'chrome-devtools-frontend.appspot.com', // Support remote Chrome DevTools frontend
   'devtools', // Support local Chrome DevTools `devtools://devtools`
 ];
 
+export const _isLocalHostname = (hostname: string) => {
+  if (hostname === 'localhost') {
+    return true;
+  }
+  let maybeIp = hostname;
+  const ipv6To4Prefix = '::ffff:';
+  if (maybeIp.startsWith(ipv6To4Prefix)) {
+    maybeIp = maybeIp.slice(ipv6To4Prefix.length);
+  }
+  if (maybeIp === '::1') {
+    return true;
+  } else if (/^127(?:.\d+){3}$/.test(maybeIp)) {
+    return maybeIp.split('.').every((part) => {
+      const num = parseInt(part, 10);
+      return num >= 0 && num <= 255;
+    });
+  } else {
+    return false;
+  }
+};
+
 export function createCorsMiddleware(exp: ExpoConfig) {
-  const allowedHostnames = [...DEFAULT_ALLOWED_CORS_HOSTNAMES];
+  const allowedHosts = [...DEFAULT_ALLOWED_CORS_HOSTS];
   // Support for expo-router API routes
   if (exp.extra?.router?.headOrigin) {
-    allowedHostnames.push(new URL(exp.extra.router.headOrigin).hostname);
+    allowedHosts.push(new URL(exp.extra.router.headOrigin).host);
   }
   if (exp.extra?.router?.origin) {
-    allowedHostnames.push(new URL(exp.extra.router.origin).hostname);
+    allowedHosts.push(new URL(exp.extra.router.origin).host);
   }
 
   return (req: ServerRequest, res: ServerResponse, next: (err?: Error) => void) => {
     if (typeof req.headers.origin === 'string') {
       const { host, hostname } = new URL(req.headers.origin);
       const isSameOrigin = host === req.headers.host;
-      if (!isSameOrigin && !allowedHostnames.includes(hostname)) {
+      const isLocalhost = _isLocalHostname(hostname);
+      const isAllowedHost = allowedHosts.includes(host) || isLocalhost;
```

Comment:
We're checking the full `host` here now, including the port, so cross-port host requests aren't automatically allowed anymore. My assumption is that this was unnecessarily permissive for `devtools://devtools`